### PR TITLE
Avoid deserializing cluster states on master

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we always serialize and compress the cluster state (or diff)
before publication and then decompress and deserialize it on the
receiving node, including on the master. This is because we perform the
(de)serialization outside of the transport service so the usual
local-node optimizations do not apply.

However on the master we then replace the received cluster state with
the local in-memory one anyway. We may as well not send it through the
transport service at all. This commit does that.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java [hunk 1]
```diff
@@ -1209,7 +1209,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     }
 
     public void testClusterRecoversAfterExceptionDuringSerialization() {
-        try (Cluster cluster = new Cluster(randomIntBetween(1, 5))) {
+        try (Cluster cluster = new Cluster(randomIntBetween(2, 5))) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java:1212]: This is a consequence of not trying to serialise anything in a one-node cluster, which weakens the (best-effort) check that we can always serialise cluster states to some extent before accepting them. I think this is ok, we don't see bugs here very much and we have tools for excising broken bits of cluster state anyway. Worth some thought, however, it's easy enough to replace this at the cost of a bit of wasted work in a one-node cluster.

**Comment by DaveCTurner** [comment]
Ugh the CI failure is legit. The voting-only-node logic requires us to go through the transport service too.

**Comment by DaveCTurner** [comment]
Ok this was supposed to be a quick and easy cleanup but is now a hack to work around the other hack in the implementation of voting-only nodes. I'm therefore leaning away from proceeding, I don't have super-strong evidence that the benefits are worth it. Leaving this open for a few days in case anyone feels strongly that we should do this.

**Comment by ywelsch** [comment]
Independently of the voting-only issue, I don't see the benefit of proceeding with this PR. It special-cases the single-master / single-node situation even more while weakening the check that the CS to be persisted in XContent is actually serializable over the wire.

**Comment by DaveCTurner** [comment]
The special single-node case isn't the main point here; the point is that we (sometimes) pointlessly decompress and deserialise a large amount of the cluster state on the master, which may be many MBs of unnecessary heap usage and a good deal of CPU too.

**Comment by DaveCTurner** [comment]
Something went OOM somewhere in the build but I don't think it was me. @elasticmachine please run elasticsearch-ci/part-1

**Comment by mark-vieira** [comment]
@elasticmachine please run elasticsearch-ci/part-1

**Comment by DaveCTurner** [comment]
> Something went OOM somewhere in the build but I don't think it was me

Turns out if you make a mistake that gets enough `CoordinatorTests` to fail then the build will OOM (we believe due to log volume). So it was me after all.

> This is therefore blocked on that issue.

Ehh actually we know about voting-only nodes in `:server` now so we can do the right thing without involving the transport interceptor. I did that.

**Comment by DaveCTurner** [comment]
Failure was a request timeout(?) talking to Homer:

```
12:35:42 Sending request to url: https://homer.app.elstc.co/webhook/jenkins/build-finished
12:36:05 Treating class org.apache.http.conn.HttpHostConnectException(Connect to homer.app.elstc.co:443 [homer.app.elstc.co/35.188.12.98] failed: Connection timed out: connect) as 408 Request Timeout
12:36:05 ERROR: Build step failed with exception
12:36:05 hudson.AbortException: Fail: Status code 408 is not in the accepted range: 100:399
```

@elasticmachine please run elasticsearch-ci/packaging-tests-windows-sample

**Review by original-brownbear — APPROVED** [review]
LGTM! Nice :)

... it's an interesting question if we could avoid all this complexity here if we had some better facilities for sending this kind of message on the transport layer (the deduplication of serialization is hard to push to it I guess ...) when we already have special casing for the bytes transport message there ... but yea just thinking out loud :)

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java [hunk 2]
```diff
@@ -368,17 +368,26 @@ public void onResponse(PublishWithJoinResponse response) {
...
 
         }
 
+        private Exception getRootCause(Exception e) {
+            if (e instanceof final TransportException transportException) {
+                if (transportException.getRootCause()instanceof final Exception rootCause) {
+                    return rootCause;
+                } else {
+                    assert false : e;
+                    logger.error(() -> new ParameterizedMessage("PublishResponseHandler: [{}] failed", discoveryNode), e);
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: Ah yeah also AFAIK just passing existing objects into a `new ParameterizedMessage` is not really any worse than capturing them in a lambda right? Anyway c00150bcba5 does as you ask.

[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: Probably not yea. It creates that `ParameterizedMessage` instance which itself creates an array of the params but I wouldn't be surprised if both compiled to the same thing in the end pretty much.

**Comment by DaveCTurner** [comment]
> it's an interesting question if we could avoid all this complexity here if we had some better facilities for sending this kind of message on the transport layer

Yeah I think it makes more sense to deduplicate the serialization and compression work at this layer where we know everything is duplicated still. I don't see a great way to tell the transport layer how to deduplicate these things. Some kind of dedicated broadcast API perhaps? It'd be pretty weird tho and ultimately would end up doing just what we do here anyway. If we had a lot of different places where we broadcast something identical and large to every node in the cluster then there'd be value in doing something more general, but I don't think we do.

Join validation is the other one that stands out in my mind, but that's also just a cluster state broadcast to all joining nodes so I reckon we could just use the `PublicationTransportHandler` for that too.

**Pull request merged**